src/components/TreeGrid/TreeGrid.tsx

```diff
@@ -152,21 +166,75 @@ export class TreeGrid extends React.PureComponent<ITreeGridProps, ITreeGridState
         return (
             <div key={key} style={style}>
                 {rowData}
-            </div>);
+            </div>
+        );
+    }
+
+    private getCheckStatus = (data: IFinalTreeNode): CheckStatus => {
+        const nodeId = this.props.treeDataSource.getIdMember(data);
+
+        if (this.props.treeDataSource.PartiallySelectedNodes[nodeId]) {
+            return CheckStatus.ChildChecked;
+        }
+
+        const selectedNodes = this.props.treeDataSource.SelectedNodes;
+        if (!selectedNodes.hasOwnProperty(nodeId)) {
+            return CheckStatus.NotChecked;
+        }
+        if (this.props.treeDataSource.SelectedNodes[nodeId]) {
+            return CheckStatus.Checked;
+        }
     }
 
     private _renderExpandCollapseButton(key, rowIndex: number, rowData: IFinalTreeNode, style, onMouseEnter, isSelectedRow: boolean) {
         const showNodeAsExpanded = rowData.isExpanded || rowData.descendantSatisfiesFilterCondition;
         let actionsTooltip = showNodeAsExpanded ? 'Collapse' : 'Expand';
-        let iconName = showNodeAsExpanded ? 'svg-icon-arrowCollapse' : 'svg-icon-arrowExpand';
+        let iconName = showNodeAsExpanded ? 'icon-arrow_down_right' : 'icon-arrow_right';
         let icon = null;
 
+        const nodeId = this.props.treeDataSource.getIdMember(rowData);
+
         if ((!rowData.children || rowData.children.length <= 0) && !rowData.hasChildren) {
             icon = null;
             actionsTooltip = null;
         } else {
-            icon = <Icon iconName={iconName} className="expand-collapse-action-icon" />;
+            icon = (
+                <Icon
+                    iconName={iconName}
+                    style={{ display: 'inline-block' }}
+                    className="expand-collapse-action-icon"
+                    onClick={(ev) => { ev.stopPropagation(); this._onTreeExpandToggleClick(ev, rowData); }}
+                />
+            );
         }
+
+        const onChange = (e) => {
+            e.preventDefault();
+            e.stopPropagation();
+
+            if (this.props.treeDataSource.SelectedNodes[nodeId]) {
+                this.props.treeDataSource.removeSelected(rowData);

```

Comment:
removeSelected and setSelected could work like the update function in TreeDataSource  before it was changed. Then you could have normal redux actions and call these functions in the reducer. If you follow the same pattern as it was with the previous version of the update function, the redux store will detect the change automatically.

Reply:
actually, disregard this, this is more internal to the treegird. You are correct that this should probably work more along the lines of isExpanded(which also just changes the state directly).